ldm/modules/attention.py

```diff
@@ -234,6 +234,7 @@ def forward(self, x, context=None):
         )
 
     def _forward(self, x, context=None):
+        x = x.contiguous()
```

Comment:
certainly this works, but I wonder whether it's premature?

the MPS `.contiguous()` workaround is only needed at the very last moment when transferring a tensor (utilising a view, e.g. a reshaped tensor) to CPU.

so if we can figure out which operation downstream of here is transferring this tensor to CPU, we could move this `.contiguous()` later (there may be performance/memory benefits).

the perf is already good of course, so we could use this for now. but would be nice to know more clearly where the problem arises and solve it a bit more surgically. could throw a TODO on there to investigate later.

Reply:
You know far more about this than I do. I'm not new to Python but I don't know anything about pytorch and AI in general. I'm taking a class now... As I've said elsewhere, I mainly got involved because I am pretty good about getting software to work on Mac, even if I don't know what it's doing... All of the changes I've done were based on what someone else did or my own guessing. In the case of the random number generator, I guessed wrong and actually broke fixed seeds using the --fixed_code flag. Someone finally figured it out but I don't know if they realized I broke it to begin with... I want to stay involved but I'm learning by watching what you guys are doing at this point.

Reply:
thanks for heads-up about `--fixed_code` 🙂 will check implications for my branch.

no worries at all. I only know about `contiguous()` because it was explained as the answer to a bug I submitted (https://github.com/pytorch/pytorch/issues/79383).  
happy for us to commit to this as step 1; @lstein confirmed it does not measurably regress performance for CUDA users.  
but since I expect optimization is possible, a TODO (e.g. linking to this comment thread) would be a good way to make sure the idea doesn't get lost.

Reply:
there's an argument that since only MPS needs this, only MPS should do it:

```python
x = x.contiguous() if x.device.type == 'mps' else x
```

the cost of `contiguous()` I think is that it has to look at the contents of the tensor through the view that's been created over it, and construct a copy of it into contiguous memory. we can easily spare CUDA and CPU users from this cost.

Reply:
Contiguous can result in the unnecessary copying of a tensor, so @Birch-san's if mps statement is a better option.

Reply:
I've modified the line according to @Birch-san 's recommendation. Tested fine.